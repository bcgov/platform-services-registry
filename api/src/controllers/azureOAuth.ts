import axios from "axios";
import { logger } from "@bcgov/common-nodejs-utils";
// import { Response } from "express";
import qs from "qs";
import { Request, Response } from "express";

export const fetchAzureAccessToken = async (req: Request, res: Response): Promise<void> => {
  try {
    const options = {
      Host: "login.microsoftonline.com",
      client_id: `${process.env.AZURE_CLIENT_ID}`,
      scope: "https://graph.microsoft.com/.default",
      client_secret: process.env.AZURE_CLIENT_SECRET,
      grant_type: "client_credentials",
    };
    const url: string = `https://login.microsoftonline.com/${process.env.AZURE_TENANT_ID}/oauth2/v2.0/token`;
    axios
      .post(url, qs.stringify(options))
      .then((response) => {
        if (!response.data.access_token) {
          throw Error(
            "a result was returned from Microsoft's token endpoint, but it doesn't seem to have an access token included"
          );
        }
        res.status(200).json(response.data.access_token);
      })
      .catch((error) => {
        logger.info(`error: ${error}`);
        res.status(500);
      });
  } catch (err) {
    logger.error(`err = ${err.message}`);
  }
};

export default fetchAzureAccessToken;
