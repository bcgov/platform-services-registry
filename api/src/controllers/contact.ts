//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-06-15.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import shared from '../libs/shared';
import { validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const createContact = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { ContactModel } = dm;

  const rv = validateObjProps(ContactModel.requiredFields, body);
  if (rv) {
    throw rv;
  }

  try {
    const result = await ContactModel.create(body);

    res.status(201).json(result);
  } catch (err) {
    const message = `Unable to create contact`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateContact = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { ContactModel } = dm;
  const { contactId } = params;
  const {
    firstName,
    lastName,
    email,
    githubId,
  } = body;

  try {
    const record = await ContactModel.findById(contactId);
    const aBody = {
      userId: record.userId,
      firstName,
      lastName,
      email,
      githubId,
      roleId: record.roleId,
    };

    const rv = validateObjProps(ContactModel.requiredFields, aBody);

    if (rv) {
      throw rv;
    }

    const results = await ContactModel.update(contactId, aBody);

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable update contact ID ${contactId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const requestContactEdit = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  try {
    res.status(201).end();
  } catch (err) {
    const message = `Unable to update quota`;
    logger.error(`${message}, err = ${err.message}`);
    throw errorWithCode(message, 500);
  }
};
