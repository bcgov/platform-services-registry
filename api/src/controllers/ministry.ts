//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { errorWithCode, logger } from "@bcgov/common-nodejs-utils";
import { Request, Response } from "express";
import DataManager from "../db";
import shared from "../libs/shared";

const dm = new DataManager(shared.pgPool);

const fetchMinistrySponsors = async (
  req: Request,
  res: Response
): Promise<void> => {
  const { MinistryModel } = dm;
  try {
    const results = await MinistryModel.findAll();

    res.status(200).json(results);
  } catch (err) {
    const message = "Unable fetch ministries";
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export default fetchMinistrySponsors;
