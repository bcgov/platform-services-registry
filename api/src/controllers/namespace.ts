//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-04-27.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import shared from '../libs/shared';
import { validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const createNamespace = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId } = params;
  const aBody = { ...body, profileId };

  const rv = validateObjProps(NamespaceModel.requiredFields, aBody);
  if (rv) {
    throw rv;
  }

  try {
    const results = await NamespaceModel.create(aBody);

    res.status(201).json(results);
  } catch (err) {
    const message = 'Unable create new namespace';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileNamespaces = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId } = params;

  try {
    const results = await NamespaceModel.findForProfile(Number(profileId));

    res.status(200).json(results);
  } catch (err) {
    const message = `Unable fetch namespace for profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileNamespace = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { namespaceId } = params;

  try {
    const results = await NamespaceModel.findById(Number(namespaceId));

    res.status(200).json(results);
  } catch (err) {
    const message = `Unable fetch namespace with ID ${namespaceId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProfileNamespace = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId, namespaceId } = params;
  const { name, clusterId } = body;
  const aBody = { name, profileId, clusterId };

  const rv = validateObjProps(NamespaceModel.requiredFields, aBody);
  if (rv) {
    throw rv;
  }

  try {
    await NamespaceModel.update(namespaceId, aBody);

    res.status(204).end();
  } catch (err) {
    const message = `Unable update project profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const archiveProfileNamespace = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId, namespaceId } = params;

  try {
    await NamespaceModel.delete(namespaceId);

    res.status(204).end();
  } catch (err) {
    const message = `Unable to archive profile ${profileId} namespace ${namespaceId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};
