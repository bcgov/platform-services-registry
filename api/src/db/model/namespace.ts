//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict';

import { logger } from '@bcgov/common-nodejs-utils';
import { Pool } from 'pg';
import { projectSetNames } from '../../constants';
import { CommonFields, Model } from './model';
import { QuotaSize } from './quota';

export interface ClusterNamespace extends CommonFields {
  namespaceId: number;
  clusterId: number;
  provisioned: boolean;
  quotaCpuSize: QuotaSize;
  quotaMemorySize: QuotaSize;
  quotaStorageSize: QuotaSize;
}

export interface ProjectNamespace extends CommonFields {
  name: string;
  profileId: number;
}

export default class NamespaceModel extends Model {
  table: string = 'namespace';
  requiredFields: string[] = [
    'name',
    'profileId',
  ];
  pool: Pool;

  constructor(pool: any) {
    super();
    this.pool = pool;
  }

  async create(data: ProjectNamespace): Promise<ProjectNamespace> {
    const query = {
      text: `
        INSERT INTO ${this.table}
          (name, profile_id)
          VALUES ($1, $2) RETURNING *;`,
      values: [
        data.name,
        data.profileId,
      ],
    };

    try {
      const results = await this.runQuery(query);
      return results.pop();
    } catch (err) {
      const message = `Unable to create namespace`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async update(namespaceId: number, data: ProjectNamespace): Promise<ProjectNamespace> {
    const values: any[] = [];
    const query = {
      text: `
        UPDATE ${this.table}
          SET
            name = $1, profile_id = $2, cluster_id = $3
          WHERE id = ${namespaceId}
          RETURNING *;`,
      values,
    };

    try {
      const record = await this.findById(namespaceId);
      const aData = { ...record, ...data };
      query.values = [
        aData.name,
        aData.profileId,
        aData.clusterId,
      ];

      const results = await this.runQuery(query);
      return results.pop();
    } catch (err) {
      const message = `Unable to update namespace ${namespaceId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async delete(namespaceId: number): Promise<ProjectNamespace> {
    const query = {
      text: `
        UPDATE ${this.table}
          SET
            archived = true
          WHERE id = ${namespaceId}
          RETURNING *;
      `,
    };

    try {
      const results = await this.runQuery(query);
      return results.pop();
    } catch (err) {
      const message = `Unable to archive namespace ${namespaceId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async createProjectSet(profileId: number, clusterId: number, prefix: string,): Promise<ProjectNamespace[]> {
    const query = {
      text: `
        INSERT INTO cluster_namespace
          (namespace_id, cluster_id, quota_cpu_size, quota_memory_size, quota_storage_size)
          VALUES ($1, $2, $3, $3, $3) RETURNING *;`,
      values: [],
    };
    try {
      const names = projectSetNames.map(n => `${prefix}-${n}`);
      const nsPromises = names.map(name => this.create({
        name,
        profileId,
      }));

      const nsResults = await Promise.all(nsPromises);
      // default quota size set to QuotaSize.Small
      const clPromises = nsResults.map(nr => this.runQuery({ ...query, values: [nr.id, clusterId, QuotaSize.Small] }));
      await Promise.all(clPromises);

      return nsResults;
    } catch (err) {
      const message = `Unable to create namespace set`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async getProjectSetProvisionStatus(profileId: number, clusterId: number): Promise<boolean> {
    const query = {
      text: `
        SELECT provisioned FROM cluster_namespace
          WHERE namespace_id = $1 AND cluster_id = $2;`,
      values: [],
    };

    try {
      const nsResults: ProjectNamespace[] = await this.findNamespacesForProfile(profileId);
      const clPromises: Promise<ClusterNamespace[]>[] = nsResults.map(nr => this.runQuery({ ...query, values: [nr.id, clusterId] }));
      const clResults: ClusterNamespace[][] = await Promise.all(clPromises);

      const clusterNamespaces = clResults.map((cl: ClusterNamespace[]) => cl.pop());
      const flags: (boolean | undefined)[] = clusterNamespaces.map((cl: (ClusterNamespace | undefined)): boolean | undefined => {
        return cl ? cl.provisioned : undefined;
      });

      if (flags.every(f => f === true)) {
        return true;
      } else if (flags.every(f => f === false)) {
        return false;
      } else {
        throw new Error(`Need to fix entries as the provisioning status of
        the project set is not consistent`);
      }
    } catch (err) {
      const message = `Unable to update provisioning status of the project set for profile ${profileId} and cluster ${clusterId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async updateProjectSetProvisionStatus(profileId: number, clusterId: number, provisioned: boolean): Promise<ProjectNamespace[]> {
    const query = {
      text: `
        UPDATE cluster_namespace
          SET provisioned = $1
          WHERE namespace_id = $2 AND cluster_id = $3
        RETURNING *;`,
      values: [],
    };

    try {
      const nsResults: ProjectNamespace[] = await this.findNamespacesForProfile(profileId);
      const clPromises: Promise<ClusterNamespace[]>[] = nsResults.map(nr => this.runQuery({ ...query, values: [provisioned, nr.id, clusterId] }));
      await Promise.all(clPromises);

      return nsResults;
    } catch (err) {
      const message = `Unable to update provisioning status of the project set for profile ${profileId} and cluster ${clusterId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async getProjectSetQuotaSize(profileId: number, clusterId: number): Promise<QuotaSize> {
    const query = {
      text: `
        SELECT quota_cpu_size, quota_memory_size, quota_storage_size FROM cluster_namespace
          WHERE namespace_id = $1 AND cluster_id = $2;`,
      values: [],
    };

    try {
      const nsResults: ProjectNamespace[] = await this.findNamespacesForProfile(profileId);
      const clPromises: Promise<ClusterNamespace[]>[] = nsResults.map(nr => this.runQuery({ ...query, values: [nr.id, clusterId] }));
      const clResults: ClusterNamespace[][] = await Promise.all(clPromises);

      const clusterNamespaces: (ClusterNamespace | undefined)[] = clResults.map(cl => cl.pop());
      const quotaSizes: QuotaSize[] = [];
      clusterNamespaces.forEach((clusterNamespace: (ClusterNamespace | undefined)): void => {
        if (!clusterNamespace) {
          return;
        }
        const { quotaCpuSize, quotaMemorySize, quotaStorageSize } = clusterNamespace;
        quotaSizes.push(quotaCpuSize, quotaMemorySize, quotaStorageSize);
      })

      const hasSameQuotaSizes: boolean = (quotaSizes.every((val, i, arr) => val === arr[0]));
      if (hasSameQuotaSizes) {
        return quotaSizes[0];
      } else {
        throw new Error(`Need to fix entries as the quota size of
        the project set is not consistent`);
      }
    } catch (err) {
      const message = `Unable to get quota size of the project set for profile ${profileId} on cluster ${clusterId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async updateProjectSetQuotaSize(profileId: number, clusterId: number, quotaSize: QuotaSize): Promise<ProjectNamespace[]> {
    const query = {
      text: `
        UPDATE cluster_namespace
          SET quota_cpu_size = $1, quota_memory_size = $1, quota_storage_size = $1
          WHERE namespace_id = $2 AND cluster_id = $3
        RETURNING *;`,
      values: [],
    };

    try {
      const nsResults = await this.findNamespacesForProfile(profileId);
      const clPromises = nsResults.map(nr => this.runQuery({ ...query, values: [quotaSize, nr.id, clusterId] }));
      await Promise.all(clPromises);

      return nsResults;
    } catch (err) {
      const message = `Unable to update quota size of the project set for profile ${profileId} on cluster ${clusterId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async findForProfile(profileId: number): Promise<ProjectNamespace[]> {
    const query = {
      text: `
        SELECT namespace.id AS "namespaceId", namespace.name,
        coalesce(
          (
            SELECT array_to_json(array_agg(row_to_json(x)))
            FROM (
              SELECT ref_cluster.id AS "clusterId", ref_cluster.name, cluster_namespace.provisioned
              FROM ref_cluster JOIN cluster_namespace ON ref_cluster.id = cluster_namespace.cluster_id
              WHERE namespace.id = cluster_namespace.namespace_id
            ) x
          ),
          '[]'
        ) AS clusters
        FROM ${this.table} WHERE namespace.profile_id = $1;
      `,
      values: [
        profileId,
      ],
    };

    try {
      return await this.runQuery(query);
    } catch (err) {
      const message = `Unable to fetch namespaces for profile ${profileId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  private async findNamespacesForProfile(profileId: number): Promise<ProjectNamespace[]> {
    const query = {
      text: `
        SELECT * FROM ${this.table}
          WHERE profile_id = $1 AND archived = false;`,
      values: [profileId],
    };

    try {
      return await this.runQuery(query);
    } catch (err) {
      const message = `Unable to find namespaces for profile ${profileId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async findClustersForProfile(profileId: number): Promise<any[]> {
    const query = {
      text: `
      SELECT DISTINCT ref_cluster.name
        FROM namespace
        JOIN cluster_namespace ON cluster_namespace.namespace_id = namespace.id
        JOIN profile ON profile.id = namespace.profile_id
        JOIN ref_cluster ON  ref_cluster.id = cluster_namespace.cluster_id
        WHERE profile.id = $1 AND profile.archived = false;`,
      values: [profileId],
    };

    try {
      return await this.runQuery(query);
    } catch (err) {
      const message = `Unable to find namespaces for profile ${profileId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }
}
