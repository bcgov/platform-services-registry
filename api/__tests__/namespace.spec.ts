//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { Response } from 'express';
import fs from 'fs';
import path from 'path';
import { Pool } from 'pg';
import { archiveProfileNamespace, createNamespace, fetchProfileNamespace, fetchProfileNamespaces, updateProfileNamespace } from '../src/controllers/namespace';

const p0 = path.join(__dirname, 'fixtures/select-namespace.json');
const select = JSON.parse(fs.readFileSync(p0, 'utf8'));

const p1 = path.join(__dirname, 'fixtures/insert-namespace.json');
const insert = JSON.parse(fs.readFileSync(p1, 'utf8'));

const client = new Pool().connect();

class FauxExpress {
  res: Partial<Response> = {
    clearCookie: jest.fn(),
    cookie: jest.fn(),
    json: jest.fn().mockImplementation((param) => {
      this.responseData = param;
      return this.res;
    }),
    status: jest.fn().mockImplementation((code) => {
      this.res.statusCode = code;
      return this.res;
    }),
    statusCode: 200,
    end: jest.fn(),
  }
  req: any;
  responseData: any;
}

describe('Namespace event handlers', () => {
  let ex;

  beforeEach(() => {
    jest.clearAllMocks();

    ex = new FauxExpress();
  });

  it('A Namespace is created', async () => {
    const req = {
      body: insert,
      params: { profileId: 1 },
    };

    const addon = {
      id: 9,
      createdAt: new Date(),
      updatedAt: new Date(),
    };

    client.query.mockReturnValueOnce({ rows: [{ ...insert, ...addon }] });

    // @ts-ignore
    await createNamespace(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot({
      id: expect.any(Number),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
    });
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });

  it('A Namespace fails to create', async () => {
    const req = {
      body: insert,
    };

    client.query.mockImplementation(() => { throw new Error() });

    // @ts-ignore
    await expect(createNamespace(req, ex.res)).rejects.toThrow();

    expect(ex.res.status).not.toBeCalled();
    expect(ex.res.json).not.toBeCalled();
  });


  it('All Namespaces are returned', async () => {
    const req = {
      params: { profileId: 1 },
    };
    client.query.mockReturnValueOnce({ rows: select });

    // @ts-ignore
    await fetchProfileNamespaces(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });

  it('Fetch all Namespaces should throw', async () => {
    const req = {
      params: {},
    };
    client.query.mockImplementation(() => { throw new Error() });

    // @ts-ignore
    await expect(fetchProfileNamespaces(req, ex.res)).rejects.toThrowErrorMatchingSnapshot();

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.responseData).toBeUndefined();
  });

  it('A single Namespace is returned', async () => {
    const req = {
      params: {
        profileId: 1,
        namespaceId: 1,
      },
    };
    client.query.mockReturnValueOnce({ rows: [select[0]] });

    // @ts-ignore
    await fetchProfileNamespace(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });

  it('Fetch single profile should throw', async () => {
    const req = {
      params: { profileId: 1 },
    };
    client.query.mockImplementation(() => { throw new Error() });

    // @ts-ignore
    await expect(fetchProfileNamespace(req, ex.res)).rejects.toThrowErrorMatchingSnapshot();

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.responseData).toBeUndefined();
  });

  it('A Namespace is updated', async () => {
    const body = JSON.parse(JSON.stringify(insert));
    // const aBody = {
    //   ...body,
    //   id: 9,
    //   createdAt: '2020-05-19T20:02:54.561Z',
    //   updateAt: '2020-05-19T20:02:54.561Z',
    // };
    const req = {
      params: {
        profileId: 1,
        namespaceId: 1,
      },
      body,
    }

    client.query.mockReturnValue({ rows: [body] });

    // @ts-ignore
    await updateProfileNamespace(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).not.toBeCalled();
  });

  it('A Namespace fails to update', async () => {
    const body = JSON.parse(JSON.stringify(insert));
    const req = {
      params: {
        profileId: 1,
      },
      body,
    }

    client.query.mockImplementation(() => { throw new Error() });

    // @ts-ignore
    await expect(updateProfileNamespace(req, ex.res)).rejects.toThrow();

    expect(ex.res.status).not.toBeCalled();
    expect(ex.res.json).not.toBeCalled();
  });

  it('A Namespace is archived', async () => {
    const req = {
      params: {
        profileId: 9,
        namespaceId: 9,
      },
    }

    client.query.mockReturnValueOnce({ rows: [{ ...insert, archived: true }] });

    // @ts-ignore
    await archiveProfileNamespace(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.json).not.toBeCalled();
    expect(ex.res.end).toBeCalled();
  });

  it('A project fails to archive', async () => {
    const req = {
      params: {
        profileId: 9,
      },
    }

    client.query.mockImplementation(() => { throw new Error() });

    // @ts-ignore
    await expect(archiveProfileNamespace(req, ex.res)).rejects.toThrow();

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.res.json).not.toBeCalled();
    expect(ex.res.end).not.toBeCalled();
  });
});
