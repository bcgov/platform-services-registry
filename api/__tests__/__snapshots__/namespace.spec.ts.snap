// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Namespace event handlers A Namespace is archived 1`] = `
Array [
  Array [
    Object {
      "text": "
        UPDATE namespace
          SET
            archived = true
          WHERE id = 9
          RETURNING *;
      ",
    },
  ],
]
`;

exports[`Namespace event handlers A Namespace is archived 2`] = `204`;

exports[`Namespace event handlers A Namespace is archived 3`] = `undefined`;

exports[`Namespace event handlers A Namespace is created 1`] = `
Array [
  Array [
    Object {
      "text": "
        INSERT INTO namespace
          (name, profile_id)
          VALUES ($1, $2) RETURNING *;",
      "values": Array [
        "abc123-test",
        1,
      ],
    },
  ],
]
`;

exports[`Namespace event handlers A Namespace is created 2`] = `201`;

exports[`Namespace event handlers A Namespace is created 3`] = `
Object {
  "clusterId": 1,
  "createdAt": Any<Date>,
  "id": Any<Number>,
  "name": "abc123-test",
  "updatedAt": Any<Date>,
}
`;

exports[`Namespace event handlers A Namespace is updated 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM namespace
          WHERE id = $1 AND archived = false;",
      "values": Array [
        1,
      ],
    },
  ],
  Array [
    Object {
      "text": "
        UPDATE namespace
          SET
            name = $1, profile_id = $2, cluster_id = $3
          WHERE id = 1
          RETURNING *;",
      "values": Array [
        "abc123-test",
        1,
        1,
      ],
    },
  ],
]
`;

exports[`Namespace event handlers A Namespace is updated 2`] = `204`;

exports[`Namespace event handlers A Namespace is updated 3`] = `undefined`;

exports[`Namespace event handlers A project fails to archive 1`] = `
Array [
  Array [
    Object {
      "text": "
        UPDATE namespace
          SET
            archived = true
          WHERE id = undefined
          RETURNING *;
      ",
    },
  ],
]
`;

exports[`Namespace event handlers A project fails to archive 2`] = `200`;

exports[`Namespace event handlers A single Namespace is returned 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM namespace
          WHERE id = $1 AND archived = false;",
      "values": Array [
        1,
      ],
    },
  ],
]
`;

exports[`Namespace event handlers A single Namespace is returned 2`] = `200`;

exports[`Namespace event handlers A single Namespace is returned 3`] = `
Object {
  "archived": false,
  "clusterId": 1,
  "createdAt": "2020-05-06T21:16:42.799Z",
  "id": 1,
  "name": "abc123-tools",
  "profileId": 4,
  "updatedAt": "2020-05-06T21:16:42.799Z",
}
`;

exports[`Namespace event handlers All Namespaces are returned 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT namespace.id AS \\"namespaceId\\", namespace.name,
        coalesce(
          (
            SELECT array_to_json(array_agg(row_to_json(x)))
            FROM (
              SELECT ref_cluster.id AS \\"clusterId\\", ref_cluster.name, cluster_namespace.provisioned
              FROM ref_cluster JOIN cluster_namespace ON ref_cluster.id = cluster_namespace.cluster_id
              WHERE namespace.id = cluster_namespace.namespace_id
            ) x
          ),
          '[]'
        ) AS clusters
        FROM namespace WHERE namespace.profile_id = $1;
      ",
      "values": Array [
        1,
      ],
    },
  ],
]
`;

exports[`Namespace event handlers All Namespaces are returned 2`] = `200`;

exports[`Namespace event handlers All Namespaces are returned 3`] = `
Array [
  Object {
    "archived": false,
    "clusterId": 1,
    "createdAt": "2020-05-06T21:16:42.799Z",
    "id": 1,
    "name": "abc123-tools",
    "profileId": 4,
    "updatedAt": "2020-05-06T21:16:42.799Z",
  },
  Object {
    "archived": false,
    "clusterId": 1,
    "createdAt": "2020-05-06T21:26:55.514Z",
    "id": 2,
    "name": "abc123-dev",
    "profileId": 4,
    "updatedAt": "2020-05-06T21:26:55.514Z",
  },
  Object {
    "archived": false,
    "clusterId": 1,
    "createdAt": "2020-05-06T21:27:00.744Z",
    "id": 3,
    "name": "abc123-test",
    "profileId": 4,
    "updatedAt": "2020-05-06T21:27:00.744Z",
  },
  Object {
    "archived": false,
    "clusterId": 1,
    "createdAt": "2020-05-06T21:27:07.148Z",
    "id": 4,
    "name": "abc123-prod",
    "profileId": 4,
    "updatedAt": "2020-05-06T21:27:07.148Z",
  },
]
`;

exports[`Namespace event handlers Fetch all Namespaces should throw 1`] = `"Unable fetch namespace for profile ID undefined"`;

exports[`Namespace event handlers Fetch all Namespaces should throw 2`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT namespace.id AS \\"namespaceId\\", namespace.name,
        coalesce(
          (
            SELECT array_to_json(array_agg(row_to_json(x)))
            FROM (
              SELECT ref_cluster.id AS \\"clusterId\\", ref_cluster.name, cluster_namespace.provisioned
              FROM ref_cluster JOIN cluster_namespace ON ref_cluster.id = cluster_namespace.cluster_id
              WHERE namespace.id = cluster_namespace.namespace_id
            ) x
          ),
          '[]'
        ) AS clusters
        FROM namespace WHERE namespace.profile_id = $1;
      ",
      "values": Array [
        NaN,
      ],
    },
  ],
]
`;

exports[`Namespace event handlers Fetch single profile should throw 1`] = `"Unable fetch namespace with ID undefined"`;

exports[`Namespace event handlers Fetch single profile should throw 2`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM namespace
          WHERE id = $1 AND archived = false;",
      "values": Array [
        NaN,
      ],
    },
  ],
]
`;
